File: Cpp/BaekJoon/Gold/G5_3079.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

int N, M;
vector<long long> timeList;

bool check(long long timeVal) {
  long long person = 0;

  for (int i = 0; i < N; i++) {
    person += timeVal / timeList[i];

    if (person >= M) return true;
  }

  return false;
}

int main() {
  cin >> N >> M;

  timeList.resize(N, 0);

  for (int i = 0; i < N; i++) {
    cin >> timeList[i];
  }

  sort(timeList.begin(), timeList.end());

  long long lo = 1;
  long long hi = timeList[N - 1] * M;

  long long ans = 0;

  while (lo <= hi) {
    long long mid = (lo + hi) / 2;

    if (check(mid)) {
      ans = mid;
      hi = mid - 1;
    } else {
      lo = mid + 1;
    }
  }

  cout << ans;

  return 0;
}

File: Java/BaekJoon/Gold/G5_3079.java
package BaekJoon.Gold;

import java.util.Arrays;
import java.util.Scanner;

public class G5_3079 {

    public static int N, M;
    public static long[] timeList = new long[100000];

    public static boolean check(long timeVal) {
        long person = 0;

        for (int i = 0; i < N; i++) {
            person += timeVal / timeList[i];

            if (person >= M) return true;
        }

        return false;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        N = sc.nextInt();
        M = sc.nextInt();

        for (int i = 0; i < N; i++) {
            timeList[i] = sc.nextLong();
        }

        Arrays.sort(timeList, 0, N);

        long lo = 1;
        long hi = timeList[N - 1] * M;

        long ans = 0;

        while(lo <= hi) {
            long mid = (lo + hi) / 2;

            if (check(mid)) {
                hi = mid - 1;
                ans = mid;
            } else {
                lo = mid + 1;
            }
        }

        System.out.println(ans);

        sc.close();
    }
}